code/graphics/color.cpp

```diff
@@ -0,0 +1,227 @@
+#include "graphics/color.h"
+#include "graphics/2d.h"
+#include "globalincs/pstypes.h"
+#include "globalincs/vmallocator.h"
+#include "math/floating.h"
+
+void hdr_color::set_vecf(const SCP_vector<float>* input)
+{
+	size_t l = input->size();
+	Assertion(l >= 3, "attempted to set color values invalid component count");
+	this->red = input->at(0);
+	this->green = input->at(1);
+	this->blue = input->at(2);
+	if (l > 3)
+		this->alpha = input->at(3);
+	if (l > 4)
+		this->intensity = input->at(4);
+}
+
+/*
+ * @brief Fills a vector with the raw float color components of the color, in the order red, green, blue, alpha,
+ * intensity
+ *
+ * @param to_fill: The fector that will be filled with the color compoents
+ */
+void hdr_color::get_v5f(SCP_vector<float>* to_fill) const
```

Comment:
You could make this a non-const reference.  In fact, this has the same pointer bug and memory risk as the function in parsehi.